### ISSUE TITLE
celery randomly throwing 500 and crashing

### ISSUE THREAD

**Issue by davetrollope** [celery/celery#4988]
We are seeing the following crash in our production logs on a frequent basis (stacktrace below). It occurs during idle times but takes down the worker.

I thought this was related to prior issue #3877 and we upgraded to 4.2.1 but we are still seeing this crash.

```
deploy@app1[production]:~$ celery report

software -> celery:4.2.1 (windowlicker) kombu:4.2.1 py:3.5.2
            billiard:3.5.0.4 py-amqp:2.3.2
platform -> system:Linux arch:64bit, ELF imp:CPython
loader   -> celery.loaders.default.Loader
settings -> transport:amqp results:disabled
```

Any advice would be greatly appreciated.

Stacktrace:

```
[CRITICAL] 2018-08-17 23:43:59,125: [worker] Unrecoverable error: Exception('Request HTTP Error  HTTP 500  Internal Server Error (b\'<?xml version="1.0"?><ErrorResponse xmlns="http://queue.amazonaws.com/doc/2012-11-05/"><Error><Type>Receiver</Type><Code>InternalError</Code><Message>We encountered an internal error. Please try again.</Message><Detail/></Error><RequestId>40b6b208-3ba4-5e38-a31e-e1f96450d403</RequestId></ErrorResponse>\')',)
Traceback (most recent call last):
  File "/var/lib/python/virtualenv/spss/lib/python3.5/site-packages/celery/worker/worker.py", line 205, in start
    self.blueprint.start(self)
  File "/var/lib/python/virtualenv/spss/lib/python3.5/site-packages/celery/bootsteps.py", line 119, in start
    step.start(parent)
  File "/var/lib/python/virtualenv/spss/lib/python3.5/site-packages/celery/bootsteps.py", line 369, in start
    return self.obj.start()
  File "/var/lib/python/virtualenv/spss/lib/python3.5/site-packages/celery/worker/consumer/consumer.py", line 317, in start
    blueprint.start(self)
  File "/var/lib/python/virtualenv/spss/lib/python3.5/site-packages/celery/bootsteps.py", line 119, in start
    step.start(parent)
  File "/var/lib/python/virtualenv/spss/lib/python3.5/site-packages/celery/worker/consumer/consumer.py", line 593, in start
    c.loop(*c.loop_args())
  File "/var/lib/python/virtualenv/spss/lib/python3.5/site-packages/celery/worker/loops.py", line 91, in asynloop
    next(loop)
  File "/var/lib/python/virtualenv/spss/lib/python3.5/site-packages/kombu/asynchronous/hub.py", line 354, in create_loop
    cb(*cbargs)
  File "/var/lib/python/virtualenv/spss/lib/python3.5/site-packages/kombu/asynchronous/http/curl.py", line 111, in on_readable
    return self._on_event(fd, _pycurl.CSELECT_IN)
  File "/var/lib/python/virtualenv/spss/lib/python3.5/site-packages/kombu/asynchronous/http/curl.py", line 124, in _on_event
    self._process_pending_requests()
  File "/var/lib/python/virtualenv/spss/lib/python3.5/site-packages/kombu/asynchronous/http/curl.py", line 130, in _process_pending_requests
    self._process(curl)
  File "/var/lib/python/virtualenv/spss/lib/python3.5/site-packages/kombu/asynchronous/http/curl.py", line 178, in _process
    buffer=buffer, effective_url=effective_url, error=error,
  File "/var/lib/python/virtualenv/spss/lib/python3.5/site-packages/vine/promises.py", line 150, in __call__
    svpending(*ca, **ck)
  File "/var/lib/python/virtualenv/spss/lib/python3.5/site-packages/vine/promises.py", line 143, in __call__
    return self.throw()
  File "/var/lib/python/virtualenv/spss/lib/python3.5/site-packages/vine/promises.py", line 140, in __call__
    retval = fun(*final_args, **final_kwargs)
  File "/var/lib/python/virtualenv/spss/lib/python3.5/site-packages/vine/funtools.py", line 100, in _transback
    return callback(ret)
  File "/var/lib/python/virtualenv/spss/lib/python3.5/site-packages/vine/promises.py", line 143, in __call__
    return self.throw()
  File "/var/lib/python/virtualenv/spss/lib/python3.5/site-packages/vine/promises.py", line 140, in __call__
    retval = fun(*final_args, **final_kwargs)
  File "/var/lib/python/virtualenv/spss/lib/python3.5/site-packages/vine/funtools.py", line 98, in _transback
    callback.throw()
  File "/var/lib/python/virtualenv/spss/lib/python3.5/site-packages/vine/funtools.py", line 96, in _transback
    ret = filter_(*args + (ret,), **kwargs)
  File "/var/lib/python/virtualenv/spss/lib/python3.5/site-packages/kombu/asynchronous/aws/connection.py", line 233, in _on_list_ready
    raise self._for_status(response, response.read())
Exception: Request HTTP Error  HTTP 500  Internal Server Error (b'<?xml version="1.0"?><ErrorResponse xmlns="http://queue.amazonaws.com/doc/2012-11-05/"><Error><Type>Receiver</Type><Code>InternalError</Code><Message>We encountered an internal error. Please try again.</Message><Detail/></Error><RequestId>40b6b208-3ba4-5e38-a31e-e1f96450d403</RequestId></ErrorResponse>')
[CRITICAL] 2018-08-17 23:43:59,624: [worker] Unrecoverable error: Exception('Request HTTP Error  HTTP 500  Internal Server Error (b\'<?xml version="1.0"?><ErrorResponse xmlns="http://queue.amazonaws.com/doc/2012-11-05/"><Error><Type>Receiver</Type><Code>InternalError</Code><Message>We encountered an internal error. Please try again.</Message><Detail/></Error><RequestId>aa1eb07c-6092-5c3a-8472-e38d3f6657d2</RequestId></ErrorResponse>\')',)
Traceback (most recent call last):
  File "/var/lib/python/virtualenv/spss/lib/python3.5/site-packages/celery/worker/worker.py", line 205, in start
    self.blueprint.start(self)
  File "/var/lib/python/virtualenv/spss/lib/python3.5/site-packages/celery/bootsteps.py", line 119, in start
    step.start(parent)
  File "/var/lib/python/virtualenv/spss/lib/python3.5/site-packages/celery/bootsteps.py", line 369, in start
    return self.obj.start()
  File "/var/lib/python/virtualenv/spss/lib/python3.5/site-packages/celery/worker/consumer/consumer.py", line 317, in start
    blueprint.start(self)
  File "/var/lib/python/virtualenv/spss/lib/python3.5/site-packages/celery/bootsteps.py", line 119, in start
    step.start(parent)
  File "/var/lib/python/virtualenv/spss/lib/python3.5/site-packages/celery/worker/consumer/consumer.py", line 593, in start
    c.loop(*c.loop_args())
  File "/var/lib/python/virtualenv/spss/lib/python3.5/site-packages/celery/worker/loops.py", line 91, in asynloop
    next(loop)
  File "/var/lib/python/virtualenv/spss/lib/python3.5/site-packages/kombu/asynchronous/hub.py", line 354, in create_loop
    cb(*cbargs)
  File "/var/lib/python/virtualenv/spss/lib/python3.5/site-packages/kombu/asynchronous/http/curl.py", line 111, in on_readable
    return self._on_event(fd, _pycurl.CSELECT_IN)
  File "/var/lib/python/virtualenv/spss/lib/python3.5/site-packages/kombu/asynchronous/http/curl.py", line 124, in _on_event
    self._process_pending_requests()
  File "/var/lib/python/virtualenv/spss/lib/python3.5/site-packages/kombu/asynchronous/http/curl.py", line 130, in _process_pending_requests
    self._process(curl)
  File "/var/lib/python/virtualenv/spss/lib/python3.5/site-packages/kombu/asynchronous/http/curl.py", line 178, in _process
    buffer=buffer, effective_url=effective_url, error=error,
  File "/var/lib/python/virtualenv/spss/lib/python3.5/site-packages/vine/promises.py", line 150, in __call__
    svpending(*ca, **ck)
  File "/var/lib/python/virtualenv/spss/lib/python3.5/site-packages/vine/promises.py", line 143, in __call__
    return self.throw()
  File "/var/lib/python/virtualenv/spss/lib/python3.5/site-packages/vine/promises.py", line 140, in __call__
    retval = fun(*final_args, **final_kwargs)
  File "/var/lib/python/virtualenv/spss/lib/python3.5/site-packages/vine/funtools.py", line 100, in _transback
    return callback(ret)
  File "/var/lib/python/virtualenv/spss/lib/python3.5/site-packages/vine/promises.py", line 143, in __call__
    return self.throw()
  File "/var/lib/python/virtualenv/spss/lib/python3.5/site-packages/vine/promises.py", line 140, in __call__
    retval = fun(*final_args, **final_kwargs)
  File "/var/lib/python/virtualenv/spss/lib/python3.5/site-packages/vine/funtools.py", line 98, in _transback
    callback.throw()
  File "/var/lib/python/virtualenv/spss/lib/python3.5/site-packages/vine/funtools.py", line 96, in _transback
    ret = filter_(*args + (ret,), **kwargs)
  File "/var/lib/python/virtualenv/spss/lib/python3.5/site-packages/kombu/asynchronous/aws/connection.py", line 233, in _on_list_ready
    raise self._for_status(response, response.read())
Exception: Request HTTP Error  HTTP 500  Internal Server Error (b'<?xml version="1.0"?><ErrorResponse xmlns="http://queue.amazonaws.com/doc/2012-11-05/"><Error><Type>Receiver</Type><Code>InternalError</Code><Message>We encountered an internal error. Please try again.</Message><Detail/></Error><RequestId>aa1eb07c-6092-5c3a-8472-e38d3f6657d2</RequestId></ErrorResponse>')
```

**Comment by auvipy** [comment]
is there any permission issue in your system?

**Comment by davetrollope** [comment]
Not that we are aware of - are there specific permissions I should check?

**Comment by thedrow** [comment]
`get_list()` performs the following operations: `ReceiveMessage` and `ListQueues`.
Your IAM needs permissions for both operations.

If you did provide the permissions for both operations, please grant *all* permissions to SQS on a staging environment and see if there is a change in behavior. If everything works, reduce the number of permissions until you can reproduce the same problem. Tedious, I know but that's what I would do. 

If that doesn't do the trick either one of two things happened:
1) AWS has a genuine error and in which case you should open a support ticket.
2) Celery is misbehaving.

I suggest that you'll open a ticket anyway if you are already paying for support.
If option 2 is the case, we need a test case.

**Comment by davetrollope** [comment]
thanks for the info, I will check those perms and dig around. since this is a random failure, its going to be tough to get a test case. Will post an update later

**Comment by davetrollope** [comment]
We verified that our permissions for the queues involved are open - users accessing the queue have wildcarded IAM permissions. Are there any specific log levels or log configurations in celery we could use to capture more output in our test env that would help diagnose this?

**Comment by thedrow** [comment]
Not really.
I checked the documentation and [InternalError ](https://docs.aws.amazon.com/AWSSimpleQueueService/latest/APIReference/CommonErrors.html)does not appear to be an error code for SQS.
Have you approached their support yet?

**Comment by thedrow** [comment]
You are correct that we should retry communicating with the queue instead of crashing.
I'm marking this as a bug.

**Comment by davetrollope** [comment]
Yes, I've opened a ticket with AWS, waiting to hear back. Will share anything useful that comes out of that conversation. Agreed, the retry will help - look forward to that being released.

**Comment by thedrow** [comment]
Feel free to contribute a fix :) PRs are always welcome.

**Comment by davetrollope** [comment]
FYI: Amazons response to my ticket:
Thank you for additional examples provided and your patience. I have finished research into your issue, and I can confirm that issues of this frequency are expected.

1. 500 errors indicate that requests to your SQS queue did not succeed, and thus are retired, this can and is known to happen when interacting with the service.

2. Please be aware that these errors are transient and can be retried. This is something that can and needs to be handled on the client side. To do this, you need to implement retries inside your applications logic with exponential backoff. Exponential backoff is an algorithmic based approach to retries that waits progressively longer each time for consecutive errors responses. I'm going to link some documentation below that will help you implement these retries for you application! [1]. Each of AWS SDK's come with the ability to implement retries with exponential backoff.

If you are continuing to experience issues where these requests do not succeed after retrying or if they are starting to become a large percentage of the total messages being processed from your queue, please get back to me with some specifics, including timestamps of the errors, the specific request ID's,
and the percentage of requests that are resulting in 500 errors. With this information we can investigate this issue further.

I would also be happy to go into more detail regarding the handling of these errors if this is required also. The logic itself does not to be too complex and the provided documentation comes with an example to help you to get started.

**Comment by jainraj** [comment]
+1

**Comment by aliaksandrmelnik1** [comment]
Almost 1 year late this bug still exists.

**Comment by thedrow** [comment]
> Almost 1 year late this bug still exists.

PRs are welcome.
We don't use SQS ourselves so we don't have a rush to fix this.

This is open source, where you are able to solve your own problems.
I promise to review any PR you provide us with.

**Comment by auvipy** [comment]
@aliaksandrmelnik1 it is easier for the people to fix this who face it in real-time.

**Comment by ebeseda** [comment]
https://github.com/celery/kombu/pull/1059

**Comment by ebeseda** [comment]
@davetrollope we have the same issue in production. It'll be fixed in 4.6.0 release
